hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/execution/ProcessorTasklet.java

```diff
@@ -246,63 +251,34 @@ private void stateMachineStep() {
                 if (pendingWatermark.equals(IDLE_MESSAGE)
                         ? outbox.offer(IDLE_MESSAGE)
                         : processor.tryProcessWatermark(pendingWatermark)) {
-                    state = PROCESS_INBOX;
+                    state = NULLARY_TRY_PROCESS;
                     pendingWatermark = null;
                 }
                 break;
 
-            case PROCESS_INBOX:
-                progTracker.notDone();
-                if (inbox.isEmpty()) {
-                    if (!isSnapshotInbox()) {
-                        if (!processor.tryProcess()) {
-                            return;
-                        }
+            case NULLARY_TRY_PROCESS:
+                if (isSnapshotInbox() || processor.tryProcess()) {
+                    state = PROCESS_INBOX;
                         outbox.reset();
-                    }
-                    fillInbox();
-                }
-                if (!inbox.isEmpty()) {
-                    if (isSnapshotInbox()) {
-                        processor.restoreFromSnapshot(inbox);
-                    } else {
-                        processor.process(currInstream.ordinal(), inbox);
-                    }
+                    stateMachineStep(); // recursion
                 }
+                break;
 
-                if (inbox.isEmpty()) {
-                    // there is either snapshot or instream is done, not both
-                    if (currInstream != null && currInstream.isDone()) {
-                        state = COMPLETE_EDGE;
-                        progTracker.madeProgress();
-                        return;
-                    } else if (numActiveOrdinals > 0
-                            && receivedBarriers.cardinality() == numActiveOrdinals) {
-                        // we have an empty inbox and received the current snapshot barrier from all active ordinals
-                        state = SAVE_SNAPSHOT;
-                        return;
-                    } else if (numActiveOrdinals == 0) {
-                        progTracker.madeProgress();
-                        state = COMPLETE;
-                    } else {
-                        state = PROCESS_WATERMARK;
-                    }
-                }
+            case PROCESS_INBOX:
+                processInbox();
                 return;
 
             case COMPLETE_EDGE:
-                progTracker.notDone();
                 if (isSnapshotInbox()
                         ? processor.finishSnapshotRestore() : processor.completeEdge(currInstream.ordinal())) {
                     assert !outbox.hasUnfinishedItem() || !isSnapshotInbox() :
                             "outbox has an unfinished item after successful finishSnapshotRestore()";
                     progTracker.madeProgress();
-                    state = initialProcessingState();
+                    state = processingState();
```

Comment:
I think it would help readability if the insides of this `processingState` method would be exposed. I'm thinking something like `state = inboundOrdinalsLeft() ? PROCESS_WATERMARK : COMPLETE`. That's something you can immediately understand without needing to look into a method body first.

Reply:
It's used 3 times, i think it makes sense to have it.